src/presign/participant.rs

```diff
@@ -489,42 +560,44 @@ impl PresignParticipant {
         info!("Presign: Handling round one message.");
 
         // First check that we have the round one public broadcast from this
-        // participant. If not, we cannot proceed, so stash that message and
-        // continue.
-        if !self
+        // participant. If not, we cannot proceed, so stash that message.
+        let r1_public_broadcast = match self
             .local_storage
-            .contains::<storage::RoundOnePublicBroadcast>(message.from())
+            .retrieve::<storage::RoundOnePublicBroadcast>(message.from())
         {
-            // Store any early round one messages.
-            info!("Presign: Stashing early round one message.");
-            self.stash_message(message)?;
-            return Ok(ProcessOutcome::Incomplete);
-        }
-        let r1_public_broadcast = self
-            .local_storage
-            .retrieve::<storage::RoundOnePublicBroadcast>(message.from())?;
+            Ok(m) => m,
+            Err(_) => {
+                // XXX This error may either be due to (1) an item not being
+                // found in storage, or (2) an internal issue with storage.
+                // Currently, this code path _assumes_ that the reason this
+                // error occurs is due to (1)!
+                info!("Presign: Stashing early round one message (no matching broadcast message).");
+                self.stash_message(message)?;
+                return Ok(ProcessOutcome::Incomplete);
+            }
+        };
 
-        let keyshare = get_keyshare(self.id, input)?;
-        let keyshare_from = input.find_auxinfo_public(message.from())?;
-        let r1_public = crate::round_one::Public::from_message(
+        let info = PresignKeyShareAndInfo::new(self.id, input)?;
+        let auxinfo_public = input.find_auxinfo_public(message.from())?;
+        crate::round_one::Public::validate_message(
```

Comment:
note: I think this is a better name but I'm going to update #214 to point out specifically that the name changed here. It looks like you didn't change the name on the round 2 or round 3 `from_message`s that also do validation, right?

Reply:
Oh yes good catch. The reasoning here was that we don't actually _need_ the message in round one, we just need to validate it. In other rounds we need the (validated) message itself, which is probably why I didn't change the names there. Probably the better name for those methods would be `retrieve_validated_message` or something similar. I can make that change here or in #214.

Reply:
That makes sense! Let's leave it for #214.